src/qvrec.1
.ig \"-*- nroff -*-
Copyright (C) 1989-1995 Free Software Foundation, Inc.

Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.

Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.

Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translations approved by the Free Software Foundation instead of in
the original English.

$Id: qvrec.1,v 1.1 2000/08/28 19:41:28 ken-ichi Exp bhepple $
$Log: qvrec.1,v $
Revision 1.1  2000/08/28 19:41:28 ken-ichi
Initial revision


..
.TH QVREC 1 "$Date: 2000/08/28 19:41:28 $" "Local" "Local commands"
.UC 4
.SH NAME
qvrec \- upload pictures to a Casio QV-10 digital camera.
.SH SYNOPSIS
.ad l
.B qvrec
[
.BI \-D \ ttydevice
]
[
.BI \-S \ speed
]
[
.B \-hv
]
file1.cam file2.cam ...
.ad b
.SH DESCRIPTION
This manual page applies to
.B qvrec
version 0.95 by ken-ichi HAYASHI and Jun-ichiro "itojun" ITOH.
.PP
.B qvrec
allows the computer to upload CAM format pictures to the Casio QV-10 digital
camera.
Use entirely at your own risk.
The CAM format is proprietary to Casio.
.P
The camera can take up to 96 pictures, each at 320x240 24-bit
pixels. It can be linked to the serial port of a computer using the
supplied cable (see
.BR qvplay (1)
for wiring changes needed for some machines).
.PP
If the camera does not respond, switch it off and then on again to do
a 'hard' reset.  
If the transfer times out, try using a slower speed.
.PP
All pictures should be 480x240 pixels.
.SH OPTIONS
Options are processed from left to right so it is important to set the
serial line and speed first.
.IP "\fB\-D\fP \fIttydevice\fP"
Specify the serial port to use. Something like \fI/dev/ttyS0\fP on 
.B Unix
and \fICOM1\fP on less sophisticated systems.
The default is \fI/dev/qvtty\fP.
.IP "\fB\-S\fP \fIspeed\fP"
Specify the baud rate for the transfer. Possible values are
{normal, mid, high, top, light} or {1,2,3,4,5} for a corresponding
{9600, 19200, 38400, 57600, 115200} baud rate.
\'light speed' is supported in Linux, OS2, and Win NT/95.
With Old Linux, use 
.B setserial(1)
to achieve lightspeed e.g.
.RS
.PP
.nf
  setserial /dev/cua1 spd_hi	# set 57600baud
  qvrec -S top file1.cam
or
  setserial /dev/cua1 spd_vhi	# set 115200baud
  qvrec -S light file2.cam
.fi
.PP
If you are using a PC/AT compatible computer, you will need a 16550A compatible
serial controller to use more than 19200 baud (the 'mid' setting).
.RE
.IP "\fB\-h\fP"
Print a help message and exit.
Report camera revision (?)
.IP "\fB\-v\fP"
Verbosely reports upload information on 
.B stderr. 
Reported information is uploaded bytes, not a filesize.
.SH ENVIRONMENT
.SH FILES
.SH BUGS
Fine mode picture of QV-100/200/300 and QV-700/770 is not supported.
.SH COPYRIGHT
.SH AUTHORS
Ken-ichi HAYASHI <xg2k-hys@asahi-net.or.jp>
.P
Jun-ichiro "itojun" Itoh <itojun@csl.sony.co.jp>
.SH AVAILABILITY
.PP
.nf
Unix systems:
-PC/AT clone Linux 2.2.X, 2.3.X
-PC/AT clone Linux 1.2.13, 1.3.20, 2.0.35
-PC/AT clone BSD/OS 2.0
-Sony NEWS-3150U2 NEWS-OS 4.2.1RN
-Sony NEWS-1860 NEWS-OS 4.2C
-Sony NEWS-4000 NEWS-OS 6.1.2
-NEC EWS4800 Rel.4.4 Rev01
-Sun Sparc Station 20 Solaris2.3(SunOS 5.3)
-Sparc clone SunOS 4.1.4
-Apple Macintosh SE/30 NetBSD 1.0
-PC/AT clone FreeBSD 2.2.6R
-PC/AT clone FreeBSD 2.0.5R(*)
-PC/AT clone FreeBSD 2.1.0R(*)
-NeXT NeXTcube(M68040-25) NEXTSTEP 3.3J(*)
.PP
.nf
Non-Unix systems:
-SHARP X68030 + Human68k 3.02
-PC/AT clone Windows NT 3.51,4.0(Intel)(work on DOS prompt)
-PC/AT clone Windows 95(work on DOS prompt)
-PC/AT clone OS/2 Warp 3.0(Intel)(work on DOS prompt)
.fi
.PP
Systems marked with (*) need the hardware hack described under
.B BUGS
abobe.
.SH SPECIAL THANKS
.PP
Bob Hepple <bhepple@pacific.net.sg> (Thanks for this man page.)
.PP
.SH "SEE ALSO"
.BR qvplay (1),
.BR qvalldel (1),
.BR cam2jpeg (1)
by Jun-ichiro "itojun" Itoh <itojun@csl.sony.co.jp>.
.PP
.BR camtoppm.pl (1)
by Shuji Senda <senda@kuis.kyoto-u.ac.jp>.
.\" For emacs:
.\" Local Variables:
.\" fill-column: 70
.\" eval:(setq filename (substring buffer-file-name (string-match "[a-zA-Z0-9_.]+$" buffer-file-name)))
.\" eval:(setq basename (substring filename 0 (string-match "\\." filename)))
.\" eval:(setq compile-command (concat "make -k " basename ".man"))
.\" End:


HISTORY
Ver ??  qvlinkという名前で開発開始。
        んでもWindows用ソフトの名前がQVLINK.EXEだった
        のを思い出して名前をqvplayに変更
        RTS問題が解決してないけど、とりあえず接続ケーブルに
        細工を行なうという前提で作成することにした。

ver.0.00            とりあえず画像データを生で吐き出すバージョン
ver.0.01(95.10.24)  itojunさんのcam2jpegの機能を組み込んで
                    いきなりjpeg吐くようにした
                    PTPC-MLで限定公開開始
ver.0.02(95.10.25) NEWS-OSで動作確認しようとしたら、
                   jpeg出力部分がおもいっきりバグってた(^^)のが
                   発覚したので修正。
                   ver 0.01送り付けられた皆様どうもすいません。
                   出勤前にちょいちょいと直したのがまずかった
                   ようです。
ver.0.03(95.10.31) itojunさんによるautoconf化（ありがとうございます）
ver.0.04(95.11.03) itojunさんによるタイミング調整、エラーリカバリ、
                   qvplayのuidじゃなくて起動したユーザのuidで
                   ファイルを出力、
                    などなどの大改良
                   (うう、原作者の技術を越えた完成度に上がってる^^)
ver.0.05(95.11.14) 接続ケーブルに工作(ハード的にRTS off)しなくても
                   動くようにある程度リトライするようにした。
                   RSPORTの設定はやっぱりMakefileではなくて
                   common.hを直すように変更
                   なぜかNEWS-OS6.1ではうまく動かなくなった(;_;)
                   NEWS-OS4.2のバイナリだと動くとゆー…謎。
                   setreuidとかは setuid、seteuidに代えました＞itojunさん
ver.0.06(95.11.21) ppm形式でのthumnail画像の書き出し機能追加
                   cam形式での画像の書き出し機能追加
                   画像の順番入れ換えも対応する予定だったが、
                   テスト中に「FATAL ERROR」のため断念。ううううう。
                   CASIOさんごめんなさいということでQV10は修理中
        (95.12.29) やっと修理完了。なんと修理を依頼した／ジマ電器
                   のオオボケ連絡ミスで１カ月もCASIOさんの修理
                   センターで忘れられてたとゆー。(--#)
ver.0.07(95.12.31) Brahmaさん(QVCOMの作者)さんからの情報をもとに
                   機能強化。
                   画像転送時のチェックサムを見るようにした。
                   メイン画像のppm形式での取り出し機能追加。
                   over 9600の高速転送機能追加。
                   やっぱり(^^)cam形式での画像の書き出しは
                   バグってたので修正。
                   setreuidがある場合はsetreuidを、ない場合はsetuid、
                   seteuidで代用するようにした。

ver.0.08(96.2.3)   -F オプションで cam を指定しても -oオプションで
                   指定した出力ファイルのファイル名が*.camにならず
                   *.jpgになってしまうバグの修正。
                   RGB(Q0)形式、BMP(24bit Windows3.x)形式のサポート。
                   X680x0(Human68k 3.0)用qvplay.xの作成。
                   Windows NT/95(Intel系)用qvplay.exeの作成。
                   -X オプション(X680x0のみ)の追加。
                   -S オプションの速度指定でlinux以外はtop, light
                   を使えないようにした。
                   usageをちょっと間違えていたのを修正。
                   環境変数QVPLAYTTYでTTYデバイスの指定が
                   できるようにした。

ver.0.09(96.3.16)  NEWS-OS 6.0で標準速以外がつかえないというバグを
                   修正。(ttyの速度を変えるとRTS/CTSの状態がリセット
                   されてしまうとは…)
                   画像データ書きもどしプログラムqvrecの作成
                   X680x0で38400baudでの通信時にtty not respondに
                   なるバグを修正(したつもり)
                   これをもって公開直前Finalβとする。
                   次で公開できるといいなぁ。
                   勝て、勝つのだ五輪代表(^^)

         (96.3.30) jpegライブラリver6 を使って ppmtocamをとりあえず
                   作ってみる。QVLINK.EXEは作成したCAM画像を表示して
                   くれたけど、QV10自身はダメ。(;_;)
                   ppmtocamはボツにすることにしてYCCで書き込む方法
                   を解析することにする。

ver.0.10pre(96.4.19)
                  qvrecでPPM形式(YCC)の画像の書き込みに対応。(^^)v
                  平野さんからの情報で、Win32版でRTSをOFFする
                  ことができるようになった。57600baudもOK。
                  PalmtopPC MLでのみ限定公開開始
        (96.4.21)
                X68k版に19200baud以上の場合は、tmiso.x等の
                通信ドライバの常駐チェックをするようにした。
                             ありがとうございます＞増田さん。
                おおぉ〜転ぶな〜、阿部典史。がんばれ〜よ〜し。v(^^)v
                Win32版もtty not respondのチェックを時刻で
                するようにした。（もちょっといい方法ないのかなあ）
                X68k版、Win32版でstdin,stdoutにsetmode()するようにした。
                VC++のプロジェクトファイルはソースファイルの場所を
                絶対パスで保存してるようなので、プロジェクトファイルを
                配布するのをやめた。う〜む…。(--;

ver.0.10(96.5.5)
                ドキュメントの整理。
                再配布ＯＫ公開バージョン。
ver.0.10(96.5.10)
                実はソースセット(qvplay-0.10.tar.gz)に
                obj/X68k/siochk.cを入れ忘れているの増田さんに
                指摘される。(^^;;;;;;;
                あわててsiochk.cを収録しなおしたものを作って
                差し替える。
                (X680x0でソースからコンパイルしない限り問題ありません。)
       (96.7.22)
                ひええええ、勝っちまったブラジルに…。v(^^)v
ver.0.10a pre1(96.8.16)
                大久保さんのOS/2パッチを取り込む
                qvrecで-hでusageを出したのちttyがおかしくなる(余計な
                close)のを修正
                qvrecでPPM形式の読み込みに失敗する場合があるのを修正
                NEXTSTEP等strdupのないOSのためにstrdupを使うのをやめた。
                Gotoさんの情報により画像プロテクト機能、４分割、９分割
                画面機能、を組み込む。さらにall delete機能のみをもつ
                qvalldelを作る
        
                友達から借りた第４次スーパーロボット対戦Ｓに
                ドップリとハマる。うう、まだシナリオの1/3までしか
                終ってない…。
        (96.8.24)どうやら巷ではQV-100が発売になったらしい(^^)
ver.0.10a pre2(96.8.26)
                marioさんの情報により電源状態のレポート機能を組み込む
                なんかACアダプタだと妙に大きな値になるんですけど…(?_?)
                電池だと安定した値が出てくるなあ

                あるかな〜と思って藤沢の電器屋を物色するが、
                QV-100はみあたらず。(^^;;
                たぶん古いリンクソフトは使えないだろうなあ。
                qvplayも画像転送系コマンド以外は使えると思うけど。
                そーいえば、ケーブルの仕様もどうなってるんだろ。
                単品売りになってるようだから、ハード仕様が変わって
                なければUnixな人はちょっと幸せかも。

                コマンド送付の後のチェックサムを確認するように
                した。確認するだけでリカバーできないけど。(^^;;;
                qvrecのcam書き込みがちょっとバグってたのを修正。
                今までバレなかったとゆーことはqvrecって誰も
                使ってないってことだな(^^;

                itojunさんの指摘によりqvrecで、check sum errorが
                出ることが発覚。どうやっても原因がわからないので
                対症療法で回避。(弱気…)

                横浜／新宿／秋葉原にいってもQV-100みつからず。全品
                改修かなぁ。これはソニーのが出るまでには市場に出ない
                かも。ソニーの買おうかなぁ(^^;;;

                X680x0用CAMビューアviewcam.xの作者のあさとさん
                からmail。さっそく試してみる。
                これでX680x0もビューアで苦労しなくてすむ。v(^^)v

      (96.9.23) 台風の翌日、とくに被害ないよな〜と思ったらGPzの
                ブレーキレバーが折れてた。(;_;) 折れたレバーが
                丁寧にシートの上においてあったから、昨日の強風の
                中作業していた電気屋さんがミスって倒したんだと思う。
                まあ天災だと思ってあきらめるしかないっす。
                なぜかレバーの予備があったので即交換。

                近くの電器屋で改装のための閉店セールをやってた。
                のぞいてみたら、Machintosh用接続キットのLK-2が
                500円で売られていたので保護する。(^^;

qvrec.c
#include "config.h"
#include "version.h"
#include <ctype.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#ifdef BINARYFILEMODE
#include <fcntl.h> /* for setmode() */
#endif
#include <stdlib.h>
#include <time.h>
#ifdef HAVE_UNISTD_H
#include <sys/types.h>
#include <unistd.h>
#endif
#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#else
#define MAXPATHLEN 256
#endif
#include "command.h"
#include "command2.h"
#include "common.h"
#ifdef X68K
#include "x68k/tty_x68.h"
#else
#ifdef _WIN32
#include "win32/getopt/getopt.h"
#include "win32/tty_w32.h"
#else
#ifdef OS2
#include "os2/tty_os2.h"
#else
#ifdef DOS
#include "dos/tty_dos.h"
#else
#include "tty.h"
#endif /* DOS */
#endif /* OS2 */
#endif /* WIN32 */
#endif /* X68 */

#include "common.h"
#include "getuint.h"
#define MAX_PICTURE_NUM_QV10 96

extern int optind, opterr;
extern char *optarg;

static int speed = 0;
static int format = PPM_P;
extern int qvverbose;
extern int qvhasvgamode;
extern int qv7xxprotocol;

static int errflg = 0;
static int all_pic_num = -1;
#ifndef DONTCAREUID
static uid_t uid, euid;
static gid_t gid, egid;
static int uidswapped = 0;
#endif

static uint8_t buf[YCC_MAXSIZ_VGA];

#define HOGE 9

void version(void) {
  static char *usagestr[] = {
      QVREC_USAGE_HEADER,
      (char *)NULL,
  };
  char **p;

  p = usagestr;
  while (*p)
    fprintf(stdout, *p++);
}

void usage(void) {
  static char *usagestr[] = {
      QVREC_USAGE_HEADER,
      "qvrec [options] filename1.cam filename2.cam ...\n",
      "\t -h           : show this usage.\n",
#if defined(__linux__) || defined(WIN32) || defined(OS2) ||                    \
    defined(__FreeBSD__) || defined(DOS)
      "\t -S speed     : serial speed. [normal middle high top light]\n",
#else
      "\t -S speed     : serial speed. [normal middle high]\n",
#endif
      "\t -V           : show version information.\n",
#ifndef X68
      "\t -D ttydevice : set tty(cua) device.\n",
#endif
      (char *)NULL,
  };
  char **p;

  p = usagestr;
  while (*p)
    fprintf(stdout, *p++);
}

void Exit(int code) {
  if (!(QVgetfd() < 0))
    QVchangespeed(DEFAULT);
  QVreset(1);
  closetty(QVgetfd());
  exit(code);
}

void put_cam(int n, char *infilename) {
  int len;
  FILE *infp;
  int i;
  int j;
  uint16_t s;
  uint16_t id;
  u_int length = 0;
  int skip = 0;
  uint8_t *u;
  uint8_t lbuf[16];
  uint8_t magic[4] = {0x07, 0x20, 0x4d, 0x4d};

  if (QVbattery() <= LOW_BATT) {
    fprintf(stderr, "LOW BATTERY, change battery or connect AC adapter.\n");
    Exit(3);
  }

  infp = stdin;
  if (infilename) {
    infp = fopen(infilename, RMODE);
    if (infp == NULL) {
      fprintf(stderr, "can't open infile(%s).\n", infilename);
      errflg++;
      goto cleanup;
    }
  }

  if (fread(lbuf, sizeof(uint8_t), 4, infp) < 0) {
    fprintf(stderr, "can't read header.\n");
    errflg++;
    goto cleanup;
  }
  if (memcmp(lbuf, magic, sizeof(magic)) != 0) {
    fprintf(stderr, "this file(%s) is not CAM format.\n", infilename);
    errflg++;
    goto cleanup;
  }
  if (fread(lbuf, sizeof(uint8_t), 2, infp) < 0) {
    fprintf(stderr, "can't read area count.\n");
    errflg++;
    goto cleanup;
  }

  s = get_uint16_t(lbuf);
  for (i = 0; i < s; i++) {
    if (fread(lbuf, sizeof(uint8_t), 16, infp) < 0) {
      fprintf(stderr, "can't read area id and length.\n");
      errflg++;
      goto cleanup;
    }

    id = get_uint16_t(lbuf);
    if (id == 3) {
      length = get_u_int(lbuf + 2);
      break;
    }
    skip = skip + (int)get_u_int(lbuf + 2);
  }

  while (skip > 0) {
    i = fread(buf, sizeof(uint8_t), ((skip > 256) ? 256 : skip), infp);
    if (i < 0) {
      fprintf(stderr, "can't skip.\n");
      errflg++;
      goto cleanup;
    }
    skip = skip - i;
  }

  u = buf;
  for (i = 0; i < (int)length; i++) {
    j = fgetc(infp);
    if (j < 0) {
      perror("fgetc");
      errflg++;
      goto cleanup;
    }
    buf[i] = (uint8_t)j;
  }

  len = QVputcam(n, length, buf);

  if (len < 0) {
    errflg++;
  }

cleanup:;
  if (infp != stdin)
    fclose(infp);

  return;
}

void put_hoge(int n, char *infilename) {
  int len;
  FILE *infp;
  int i;
  int j;
  uint16_t s;
  uint16_t id;
  u_int length = 0;
  int skip = 0;
  uint8_t *u;
  uint8_t magic[4] = {0xff, 0xd8, 0xff, 0xe0};

  if (QVbattery() <= LOW_BATT) {
    fprintf(stderr, "LOW BATTERY, change battery or connect AC adapter.\n");
    Exit(3);
  }

  infp = stdin;
  if (infilename) {
    infp = fopen(infilename, RMODE);
    if (infp == NULL) {
      fprintf(stderr, "can't open infile(%s).\n", infilename);
      errflg++;
      goto cleanup;
    }
  }

  u = buf;
  length = 0;
  while (1) {
    j = fgetc(infp);
    if (j < 0) {
      fprintf(stderr, "length = %d\n", length);
      /*      perror("fgetc"); */
      goto fooo;
    }
    buf[i] = (uint8_t)j;
    length++;
  }
fooo:
  len = QVputJpegQv7xx(n, length, buf);
  fprintf(stderr, "jpeg7xx = %d\n", length);
  fprintf(stderr, "len = %d\n", len);

  if (len < 0) {
    errflg++;
  }

cleanup:;
  if (infp != stdin)
    fclose(infp);

  return;
}

void skip_space(infp) FILE *infp;
{
  int c;
  c = ' ';
  while (isspace(c)) {
    c = fgetc(infp);
  }
  if (c == EOF)
    return;
  ungetc(c, infp);
}

int fgetline(s, size, infp)
char *s;
int size;
FILE *infp;
{
  int i;
  int c;
fgetretry:
  skip_space(infp);
  for (i = 0; i < size; i++) {
    c = fgetc(infp);
    if (c == EOF)
      return (EOF);
    if (isspace(c)) {
      s[i] = '\0';
      if (c != '\n')
        skip_space(infp);
      return (i);
    } else if (c == '#') {
      while (fgetc(infp) != '\n')
        ;
      s[i] = '\0';
      if (i == 0)
        goto fgetretry;
      return (i);
    } else
      s[i] = c;
  }
  return (i);
}

#define NORM(x)                                                                \
  {                                                                            \
    if (x < 0)                                                                 \
      x = 0;                                                                   \
    else if (x > 255)                                                          \
      x = 255;                                                                 \
  }
void put_ppm(int n, char *infilename, int skiphead) {
  int len;
  FILE *infp;

  int x, y;
  uint8_t *Y;
  uint8_t *Cr;
  uint8_t *Cb;
  int r, g, b;
  long L;
  long cr, cb;
  int i, j;

  uint8_t lbuf[256];
  uint8_t magic[2] = {'P', '6'};

  int R[PICTURE_WIDTH / 3];
  int G[PICTURE_WIDTH / 3];
  int B[PICTURE_WIDTH / 3];

  if (QVbattery() <= LOW_BATT) {
    fprintf(stderr, "LOW BATTERY, change battery or connect AC adapter.\n");
    Exit(3);
  }

  infp = stdin;
  if (infilename) {
    infp = fopen(infilename, RMODE);
    if (infp == NULL) {
      fprintf(stderr, "can't open infile(%s).\n", infilename);
      errflg++;
      goto cleanup;
    }
  }

  if (!skiphead) {
    if (fgetline(lbuf, 3, infp) < 0) {
      fprintf(stderr, "can't read header.\n");
      errflg++;
      goto cleanup;
    }

    if (memcmp(lbuf, magic, sizeof(magic)) != 0) {
      fprintf(stderr, "this file(%s) is not PPM(P6) format.\n", infilename);
      errflg++;
      goto cleanup;
    }

    if (fgetline(lbuf, sizeof(lbuf), infp) < 0) {
      fprintf(stderr, "can't read width data.\n");
      errflg++;
      goto cleanup;
    }

    if (strcmp((char *)lbuf, "480") != 0) {
      fprintf(stderr, "this picture's width is not 480.\n");
      errflg++;
      goto cleanup;
    }

    if (fgetline(lbuf, sizeof(lbuf), infp) < 0) {
      fprintf(stderr, "can't read height data.\n");
      errflg++;
      goto cleanup;
    }

    if (strcmp((char *)lbuf, "240") != 0) {
      fprintf(stderr, "this picture's width is not 240.\n");
      errflg++;
      goto cleanup;
    }

    if (fgetline(lbuf, sizeof(lbuf), infp) < 0) {
      fprintf(stderr, "can't read color depth data.\n");
      errflg++;
      goto cleanup;
    }

    if (strcmp((char *)lbuf, "255") != 0) {
      fprintf(stderr, "this picture's depth is not 255.\n");
      errflg++;
      goto cleanup;
    }
  }

  Y = buf;
  Cb = Y + (PICTURE_WIDTH * PICTURE_HEIGHT);
  Cr = Cb + (PICTURE_WIDTH / 3 * PICTURE_HEIGHT / 2);

  memset(R, 0, sizeof(R));
  memset(G, 0, sizeof(G));
  memset(B, 0, sizeof(B));

  j = 0;
  for (y = 0; y < PICTURE_HEIGHT; y++) {
    for (x = 0; x < PICTURE_WIDTH; x++) {
      r = fgetc(infp);
      g = fgetc(infp);
      b = fgetc(infp);
      if ((r < 0) || (g < 0) || (b < 0)) {
        perror("fgetc");
        errflg++;
        goto cleanup;
      }

      L = (r * 29900 + g * 58700 + b * 11400) / 100000;
      NORM(L);
      Y[PICTURE_WIDTH * y + x] = (unsigned char)L;

      R[x / 3] = R[x / 3] + r;
      G[x / 3] = G[x / 3] + g;
      B[x / 3] = B[x / 3] + b;
    }
    if (y % 2 == 1) {
      for (i = 0; i < PICTURE_WIDTH / 3; i++) {
        r = R[i] / 6;
        g = G[i] / 6;
        b = B[i] / 6;

        cb = (50000 * b - 16874 * r - 33126 * g) / 100000;
        if (cb < 0)
          cb = cb + 256;
        NORM(cb);
        cr = (50000 * r - 41869 * g - 8131 * b) / 100000;
        if (cr < 0)
          cr = cr + 256;
        NORM(cr);
        Cb[PICTURE_WIDTH / 3 * j + i] = (unsigned char)cb;
        Cr[PICTURE_WIDTH / 3 * j + i] = (unsigned char)cr;
        R[i] = 0;
        G[i] = 0;
        B[i] = 0;
      }
      j++;
    }
  }

  len = QVputpicture(n,
                     PICTURE_WIDTH * PICTURE_HEIGHT +
                         PICTURE_WIDTH / 3 * PICTURE_HEIGHT / 2 * 2,
                     buf);

  if (len < 0) {
    errflg++;
  }

cleanup:;
  if (infp != stdin)
    fclose(infp);

  return;
}

#ifndef DONTCAREUID
void daemonuid(void) {
  if (uidswapped) {
#ifdef HAVE_SETREUID
    setreuid(uid, euid);
    setregid(gid, egid);
#else
    setuid(uid);
    seteuid(euid);
    setgid(gid);
    setegid(egid);
#endif
    uidswapped = 0;
  }
}

void useruid(void) {
  if (!uidswapped) {
#ifdef HAVE_SETREUID
    setregid(egid, gid);
    setreuid(euid, uid);
#else
    setgid(egid);
    setegid(gid);
    setuid(euid);
    seteuid(uid);
#endif
    uidswapped = 1;
  }
}
#endif

int main(int argc, char **argv) {
  char *devpath = NULL;
  char *outfilename = NULL;
  char c;
  qvverbose = 0;
  qvhasvgamode = 0;
  qv7xxprotocol = 0;

#ifndef DONTCAREUID
  uid = getuid();
  euid = geteuid();
  gid = getgid();
  egid = getegid();
  useruid();
#endif

  devpath = getenv("QVPLAYTTY");

  while ((c = getopt(argc, argv, "D:S:F:hvV")) != -1) {
    switch (c) {
    case 'V':
      usage();
      exit(0);
    case 'h':
    case '?':
      usage();
      exit(0);
    case 'v':
      qvverbose = 1;
      break;
    case 'D':
      devpath = optarg;
      break;
    case 'S':
      switch (optarg[0]) {
#if defined(__linux__) || defined(WIN32) || defined(OS2) ||                    \
    defined(__FreeBSD__) || defined(DOS)
      case 'l':
      case '5':
        speed = LIGHT;
        break;
      case 't':
      case '4':
        speed = TOP;
        break;
#endif
      case 'h':
      case '3':
        speed = HIGH;
        break;
      case 'm':
      case '2':
        speed = MID;
        break;
      default:
        speed = DEFAULT;
        break;
      }
      break;
    case 'F':
      switch (optarg[0]) {

      case 'h':
        format = HOGE;
        break;

      case 'C':
      case 'c':
        format = CAM;
        break;
#ifndef USEWORKFILE
      case 'R':
      case 'r':
        format = RGB_P;
        break;
      default:
        format = PPM_P;
        break;
#else
      default:
        format = CAM;
#endif
      }
      break;
    }
  }

  if (devpath == NULL) {
    devpath = malloc(sizeof(char) * (strlen(RSPORT) + 1));
    if (devpath == NULL) {
      fprintf(stderr, "can't malloc\n");
      exit(1);
    }
    strcpy(devpath, RSPORT);
  }

  if (devpath) {
#ifndef DONTCAREUID
    daemonuid();
#endif
    QVsetfd(opentty(devpath));
#ifndef DONTCAREUID
    useruid();
#endif
    if (QVgetfd() < 0)
      Exit(1);
    if (all_pic_num < 0)
      all_pic_num = QVhowmany();
    if (all_pic_num < 0)
      Exit(1);
  }

  qvhasvgamode = QVrevision() & 0x01000000;
  if ((QVrevision() & 0x00a00000) == 0x00a00000)
    qv7xxprotocol = 1;
  if (qv7xxprotocol != 0) {
    QVnewprotocol();
    fprintf(stderr, "NEW protocol \n");
  }
  QVchangespeed(speed);
  if (optind == argc) {
    if (all_pic_num < 0)
      Exit(1);
    if (qvhasvgamode) {
      if (QVremain(1) < 0) {
        fprintf(stderr, "picture full.\n");
        errflg++;
        return 1;
      }
      if (QVremain(0) < 0) {
        fprintf(stderr, "picture full.\n");
        errflg++;
        return 1;
      }
    } else {
      if (all_pic_num >= MAX_PICTURE_NUM_QV10) {
        fprintf(stderr, "MEMORY FULL, QV-10 has %d pictures.\n",
                MAX_PICTURE_NUM_QV10);
        Exit(2);
      }
    }
#ifdef BINARYFILEMODE
#ifdef _WIN32
    _setmode(_fileno(stdin), _O_BINARY);
#else
    setmode(fileno(stdin), O_BINARY);
#endif
#endif
    switch (format) {
    case CAM:
      put_cam(all_pic_num + 1, NULL);
      break;
    case RGB_P:
      put_ppm(all_pic_num + 1, NULL, 1);
      break;
    case PPM_P:
    default:
      put_ppm(all_pic_num + 1, NULL, 0);
      break;
    }
  }
  if (optind < argc) {
    while (optind < argc) {
      all_pic_num = QVhowmany();
      if (all_pic_num < 0)
        Exit(1);
      if (qvhasvgamode) {
        if (QVremain(1) < 0) {
          fprintf(stderr, "picture full.\n");
          errflg++;
          return 1;
        }
        if (QVremain(0) < 0) {
          fprintf(stderr, "picture full.\n");
          errflg++;
          return 1;
        }
      } else {
        if (all_pic_num >= MAX_PICTURE_NUM_QV10) {
          fprintf(stderr, "NO MEMORY, QV-10 has %d pictures.\n",
                  MAX_PICTURE_NUM_QV10);
          Exit(2);
        }
      }
      switch (format) {

      case HOGE:
        put_hoge(all_pic_num + 1, argv[optind++]);
        break;

      case CAM:
        put_cam(all_pic_num + 1, argv[optind++]);
        break;
      case RGB_P:
        put_ppm(all_pic_num + 1, argv[optind++], 1);
        break;
      case PPM_P:
      default:
        put_ppm(all_pic_num + 1, argv[optind++], 0);
        break;
      }
      QVshowpicture(all_pic_num + 1);
    }
  }

  Exit(errflg ? 1 : 0);
}
